cylc/flow/task_hold_mgr.py

```diff
@@ -0,0 +1,171 @@
+
+
+class PseudoTaskProxy:
+    """Used to present a pseudo task pool, for task filtering."""
+
+    def __init__(
+        self,
+        name: str,
+        point: 'PointBase',
+        ancestors: List[str],
+        state: str = TASK_STATUS_WAITING
+    ):
+        self.point = point
+        self.tdef = self.PseudoTaskdef(name, ancestors)
+        self.state = self.PseudoTaskState(state)
+
+    class PseudoTaskState:
+        def __init__(self, state):
+            self.status = state
+
+    class PseudoTaskdef:
+        def __init__(self, name: str, ancestors: List[str]):
+            self.name = name
+            self.namespace_hierarchy = ancestors
+
+
+class TaskHoldMgr:
+    """Central logic for active and future task hold/release management."""
+
+    def __init__(
+        self,
+        workflow_db_mgr: 'WorkflowDatabaseManager',
+        data_store_mgr: 'DataStoreMgr',
+        ancestors: Dict[str, List[str]]
+    ) -> None:
+        self.ancestors = ancestors
+        self.workflow_db_mgr = workflow_db_mgr
+        self.data_store_mgr = data_store_mgr
+        self.store: Set[Tuple[str, 'PointBase']] = set()
+
+    def load_from_db(self):
+        """Update the task hold store from the database."""
+        self.store.update(
+            (name, get_point(cycle)) for name, cycle in
+            self.workflow_db_mgr.pri_dao.select_tasks_to_hold()
+        )
+
+    def is_held(self, name: str, point: 'PointBase') -> bool:
+        """Is this task listed in the hold store?"""
+        return (name, point) in self.store
+
+    def hold_active_tasks(self, itasks: List['TaskProxy']) -> None:
+        """Hold an active task and add it to the hold store."""
+        for itask in itasks:
+            itask.state_reset(is_held=True)
+            self.store.add((itask.tdef.name, itask.point))
+            self.data_store_mgr.delta_task_held(itask)
+            self.workflow_db_mgr.put_tasks_to_hold(self.store)
+
+    def hold_future_tasks(
+        self, tasks: Set[Tuple[str, 'PointBase']]
+    ) -> None:
+        """Add a future task to the hold store."""
+        for name, cycle in tasks:
+            self.data_store_mgr.delta_task_held((name, cycle, True))
+        self.store.update(tasks)
+        self.workflow_db_mgr.put_tasks_to_hold(self.store)
+
+    def remove_active_task(
+        self, itask: 'TaskProxy'
+    ) -> None:
+        """Remove (not release) an active task from the hold store."""
+        self.store.discard((itask.tdef.name, itask.point))
+        self.workflow_db_mgr.put_tasks_to_hold(self.store)
+
+    def release_future_tasks(
+        self, ftasks: List['PseudoTaskProxy']
+    ) -> None:
+        """Release future tasks from the hold store."""
+        matched = set()
+        for ftask in ftasks:
+            name = ftask.tdef.name
+            cycle = ftask.point
+            matched.add((ftask.tdef.name, ftask.point))
+            self.data_store_mgr.delta_task_held((name, cycle, False))
```

Comment:
Ah, nice, so this is why they now appear released at n>0:
![image](https://github.com/cylc/cylc-flow/assets/11400777/80be6a16-bb49-47df-999a-71d096ad30c9)